sia/src/rhp/quic.rs

```diff
@@ -0,0 +1,340 @@
+        single-stack instead since IPv4 is the preferred fallback.
+    */
+    endpoint_v4: Option<Endpoint>,
+    endpoint_v6: Option<Endpoint>,
+
+    hosts: Mutex<HashMap<PublicKey, Vec<NetAddress>>>,
+    open_conns: Mutex<HashMap<PublicKey, Connection>>,
+    cached_prices: Mutex<HashMap<PublicKey, HostPrices>>,
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("rhp error: {0}")]
+    RHP(#[from] RHPError),
+
+    #[error("unknown host: {0}")]
+    UnknownHost(PublicKey),
+
+    #[error("invalid port: {0}")]
+    InvalidPort(#[from] ParseIntError),
+
+    #[error("invalid address: {0}")]
+    InvalidAddress(String),
+
+    #[error("I/O error: {0}")]
+    Io(#[from] std::io::Error),
+
+    #[error("invalid prices")]
+    InvalidPrices,
+
+    #[error("invalid signature")]
+    InvalidSignature,
+
+    #[error("connection error: {0}")]
+    Connection(String),
+}
+
+impl Dialer {
+    pub fn new(mut client_config: ClientConfig) -> Result<Self, &'static str> {
+        client_config.alpn_protocols = vec![b"sia/rhp4".to_vec()];
+
+        let client_config = QuicClientConfig::try_from(client_config).unwrap();
+        let client_config = quinn::ClientConfig::new(Arc::new(client_config));
+
+        let endpoint_v4 = match quinn::Endpoint::client((Ipv4Addr::UNSPECIFIED, 0).into()) {
+            Ok(mut endpoint) => {
+                endpoint.set_default_client_config(client_config.clone());
+                Some(endpoint)
+            }
+            Err(_) => None,
+        };
+        let endpoint_v6 = match quinn::Endpoint::client((Ipv6Addr::UNSPECIFIED, 0).into()) {
+            Ok(mut endpoint) => {
+                endpoint.set_default_client_config(client_config);
+                Some(endpoint)
+            }
+            Err(_) => None,
+        };
+
+        if endpoint_v4.is_none() && endpoint_v6.is_none() {
+            return Err("unable to create IPv4 and IPv6 endpoint");
+        }
+
+        Ok(Self {
+            endpoint_v4,
+            endpoint_v6,
+            hosts: Mutex::new(HashMap::new()),
+            open_conns: Mutex::new(HashMap::new()),
+            cached_prices: Mutex::new(HashMap::new()),
+        })
+    }
+
+    fn get_cached_prices(&self, host_key: &PublicKey) -> Option<HostPrices> {
+        let mut cached_prices = self.cached_prices.lock().unwrap();
+        match cached_prices.get(host_key) {
+            Some(prices) => {
+                if prices.valid_until < OffsetDateTime::now_utc() {
+                    cached_prices.remove(host_key);
+                    None
+                } else {
+                    Some(prices.clone())
+                }
+            }
+            _ => None,
+        }
+    }
+
+    fn set_cached_prices(&self, host_key: &PublicKey, prices: HostPrices) {
+        self.cached_prices.lock().unwrap().insert(*host_key, prices);
+    }
+
+    async fn with_host<T, F, C>(&mut self, host: PublicKey, cb: C) -> Result<T, Error>
```

Comment:
I don't know if this is a good pattern in Rust due to the borrow checker probably getting easily annoyed if the callback references anything needed in here. Often times what you see instead is a helper struct being returned like maybe an `ActiveStream` that performs any cleanup such as closing the stream or removing the host when it goes out of scope or upon `drop`.

You probably need to specify a lifetime for `self` to indicate that `ActiveStream` never outlives `self` but apart from that it might be nice.

Reply:
You would also need to copy the map, which would not be nice